File: pine-scripts/python/rsi-multi.js
// Three.js equivalent of the Pine Script RSI Multi-View indicator

// 1. --- Inputs & Enums ---
const ChartType = {
    candles: "Standard RSI Candles",
    ha: "Heikin Ashi RSI",
    line: "Standard RSI Line"
};

let chartMode = ChartType.candles; // Display Mode
let rsiLen = 14; // RSI Length
let upColor = 0x008080; // Bullish Color (Teal)
let downColor = 0x800000; // Bearish Color (Maroon)

// 2. --- RSI OHLC Calculations ---
// We calculate RSI for each component of a price bar
let rO = calculateRSI(openPrices, rsiLen);
let rH = calculateRSI(highPrices, rsiLen);
let rL = calculateRSI(lowPrices, rsiLen);
let rC = calculateRSI(closePrices, rsiLen);

// 3. --- Heikin Ashi RSI Calculations ---
let haO = null;
let haC = (rO + rH + rL + rC) / 4;
// HA Open = (Previous HA Open + Previous HA Close) / 2
haO = haO === null ? (rO + rC) / 2 : (haO + haC) / 2;
let haH = Math.max(rH, Math.max(haO, haC));
let haL = Math.min(rL, Math.min(haO, haC));

// 4. --- Selection Logic ---
let isLine = chartMode === ChartType.line;
let isHA = chartMode === ChartType.ha;

// Use HA values if selected, otherwise Standard RSI Candle values
let plotO = isHA ? haO : rO;
let plotH = isHA ? haH : rH;
let plotL = isHA ? haL : rL;
let plotC = isHA ? haC : rC;

// Determine Candle Color
let barColor = plotC >= plotO ? upColor : downColor;

// 5. --- Plotting ---
// Plotting logic for Three.js would go here
if (!isLine) {
    plotCandle(plotO, plotH, plotL, plotC, barColor);
}

// Plot a standard line if Line mode is selected
if (isLine) {
    plotLine(rC, 0x808080); // Gray line
}

// Reference Levels update
drawHorizontalLine(70, 0xFF0000, 'Overbought'); // Red dashed line
drawHorizontalLine(50, 0x808080, 'Mean'); // Gray dotted line
drawHorizontalLine(30, 0x00FF00, 'Oversold'); // Green dashed line

File: pine-scripts/python/rsi-multi.py
# @version=6
# indicator("RSI Multi-View: Candles, HA & Line", shorttitle="RSI Multi", overlay=false)

# 1. --- Inputs & Enums ---
# enum ChartType
#     candles = "Standard RSI Candles"
#     ha      = "Heikin Ashi RSI"
#     line    = "Standard RSI Line"

chartMode = input.enum(ChartType.candles, "Display Mode", group="Visuals")
rsiLen    = input.int(14, "RSI Length", minval=1, group="Settings")
upColor   = input.color(color.teal, "Bullish Color", group="Colors")
downColor = input.color(color.maroon, "Bearish Color", group="Colors")

# 2. --- RSI OHLC Calculations ---
# We calculate RSI for each component of a price bar
rO = ta.rsi(open, rsiLen)
rH = ta.rsi(high, rsiLen)
rL = ta.rsi(low, rsiLen)
rC = ta.rsi(close, rsiLen)

# 3. --- Heikin Ashi RSI Calculations ---
haO = None
haC = (rO + rH + rL + rC) / 4
# HA Open = (Previous HA Open + Previous HA Close) / 2
haO = (haO + haC) / 2 if haO is not None else (rO + rC) / 2
haH = max(rH, haO, haC)
haL = min(rL, haO, haC)

# 4. --- Selection Logic ---
isLine = chartMode == ChartType.line
isHA   = chartMode == ChartType.ha

# Use HA values if selected, otherwise Standard RSI Candle values
plotO = haO if isHA else rO
plotH = haH if isHA else rH
plotL = haL if isHA else rL
plotC = haC if isHA else rC

# Determine Candle Color
barColor = upColor if plotC >= plotO else downColor

# 5. --- Plotting ---
# plotcandle() returns 'na' if isLine is true, hiding the candles
plotcandle(None if isLine else plotO, plotH, plotL, plotC, "RSI Candles", color=barColor, wickcolor=barColor, bordercolor=barColor)

# Plot a standard line if Line mode is selected
plot(rC if isLine else None, "RSI Line", color=color.gray, linewidth=2)

# Reference Levels update
hline(70, "Overbought", color=color.new(color.red, 60), linestyle=hline.style_dashed)
hline(50, "Mean", color=color.new(color.gray, 60), linestyle=hline.style_dotted)
hline(30, "Oversold", color=color.new(color.green, 60), linestyle=hline.style_dashed)
